mteb/evaluation/evaluators/RerankingEvaluator.py

```diff
@@ -251,3 +253,165 @@ def ap_score(is_relevant, pred_scores):
         # ap = np.mean([np.mean(preds[: k + 1]) for k in range(len(preds)) if preds[k]])
         ap = average_precision_score(is_relevant, pred_scores)
         return ap
+
+
+class MIRACLRerankingEvaluator(RerankingEvaluator):
```

Comment:
@orionw and @Muennighoff if you guys have the time to look at this I would love a second look.

Reply:
It might be worth considering only keeping one evaluator (while being aware that this change it breaking, the switch from MTEB --> MMTEB might be the right time to do it)

Reply:
(If we choose to replace it I have a few more comments here to make the evaluator easier to maintain in the future)

Reply:
Looking at the changes, it seems like the reason for a new evaluator class is the `candidates` key instead of `negatives`?  I agree that it would be nice to keep only one `RerankingEvaluator` if it's possible to massage the data into the same format. 

I could be missing something though, @crystina-z is there something I'm missing about the MiRACL data that makes the normal evaluator not work for it? Or would it be possible to rename the keys so that it could fit the same format? 

Thanks for the PR, very excited to see MiRACL added!

Reply:
heyy! I won't say `candidates` is a replacement for `negatives`, so previously reranking evaluation always includes all the `positives`, whereas it might not be true for reranking: The top-k `candidates` might contain all or partial or even zero positive passages. That's my main motivation to add a separate evaluator to rerank precisely the top-k from the previous stage. But lmk how you think!

Reply:
Makes sense, so `candidates` in your evaluator is `positives` + `negatives` in the original evaluator i.e. more or less the same, no? I also think it'd be great to unify the two.

Another small difference I think are the metrics you are calculating (ndcg etc.) but I think it'd be fine to add them to the original evaluator and calculate more metrics there for all reranking evals in the future if those metrics are important to you.

Reply:
yup `candidates` functions the same with `positives` + `negatives`, tho it doesn't necessarily include all `positives`

Reply:
Those missing positives can just be filtered out, no? It seems like they wouldn't affect the score, would they? Then maybe your dataet can be turned into the same format as other Reranking tasks and we can use the original evaluator by adding your desired metrics to it.

Reply:
oh I believe they would affect the scores? Say among the 100 queries, there are 20 queries with no positive examples returned from the previous stage, then the final reranking score would be divided by all 100 rather than just the 80 queries that have positive documents from the previous stages.

Reply:
I think I may be misunderstanding still - I thought most reranking tasks had some positive (but not all possible ones in all cases) and then the negatives.  The only way to have all the positives present in the Top-K always is if the task was only designed for reranking. 

I think in the MIRACL case it makes sense to include the top 100 (or N) from a standard first-stage model (this seems like BM25, which is great) and then rerank them, even if all positives are not included.  Then the numbers will be realistic as if one had a first stage model and will match other reported numbers on MIRACL reranking. I think this also aligns with other reranking tasks in MTEB, like the SciDocs one, and we can still use the same evaluator.  

Is this the current plan? It might be this is what y'all meant above and I'm just catching up

Reply:
> Then the numbers will be realistic as if one had a first-stage model and will match other reported numbers on MIRACL reranking.

Yup that's the major concern of us. It'd be nice to have an evaluation aligned with the reranking results outside of mmteb.

Alternatively, do you think we can have multiple evaluator for a task? say an "all-positive-included evaluator" (current RerankingEvaluator) and a "first-stage-only evaluator" (the current MIRACLRerankingEvaluator), and users and specify which evaluators they wanna use. I can update the `mmteb-miracl-reranking` hgf dataset so that it supports both.

Reply:
> say an "all-positive-included evaluator" (current RerankingEvaluator) 

Ah I think I miscommunicated - the current `RerankingEvaluator` does this already. You just have to set the positives to contain only the positives included in the top-K from BM25.  This is what the current [SciDocsReranking](https://github.com/embeddings-benchmark/mteb/blob/main/mteb/tasks/Reranking/eng/SciDocsReranking.py) is setup to do AFAIK.

The `RerankingEvaluator` is agnostic from what the set of positives are -- so each task can define what they want the set of positives to be. You've set it to be BM25 which I think is a great and thus, there's no need for positives to contain all the positives in the qrels for reranking.

Reply:
ah right, but the queries without any positives would be skipped during evaluation or that could be bypassed as well?

Reply:
Are there queries without positives in the top 100? I'm surprised by that, must be good queries :) 

If we don't include those queries in MTEB I think the eval for MTEB will have a higher score than other rerankers evaluated on MIRACL, so maybe we could keep them just to keep the score comparable? I assume there aren't many queries like this so there won't be too much wasted compute, but could be wrong.

I'm not sure if the current MTEB setup supports positives being an empty list but we can add that to MTEB if we want to keep the eval the same inside MTEB and out

Reply:
yea supporting positives as an empty list would be nice! I guess we don't even need to really run reranking on them, but just consider those queries when computing the eval metrics (do average on all queries).

I also just quickly summed up the number and percentage of queries that has no positive in top-100 BM25, to get a sense of how much it would impact the scores. It could be non-trivial for some languages (e.g., `zh` has around 30%)

ar: 187 out of 2896 queries (6.46%)
bn: 21 out of 411 queries (5.11%)
en: 82 out of 799 queries (10.26%)
es: 72 out of 648 queries (11.11%)
fa: 117 out of 632 queries (18.51%)
fi: 70 out of 1271 queries (5.51%)
fr: 74 out of 343 queries (21.57%)
hi: 31 out of 350 queries (8.86%)
id: 31 out of 960 queries (3.23%)
ja: 108 out of 860 queries (12.56%)
ko: 27 out of 213 queries (12.68%)
ru: 250 out of 1252 queries (19.97%)
sw: 106 out of 482 queries (21.99%)
te: 138 out of 828 queries (16.67%)
th: 60 out of 733 queries (8.19%)
zh: 121 out of 393 queries (30.79%)
de: 83 out of 305 queries (27.21%)
yo: 30 out of 119 queries (25.21%)